### PR TITLE
Converting Methods

### PR DESCRIPTION
**Pull request by dlid**
Hi.

In a project we needed to convert C# interfaces into TypeScript, and one crucial part of this was that the methods of the interfaces were incldued as well..

I really liked the look of this project of yours and I took the liberty to add some basic functionality to include the methods as well.

- The methods are included in both classes and interfaces
- I added the new boolean option `methods` (default: false) that can be set to *true* to include methods.
- I added the new boolean option `returnPromise` (default false) to return the type as a promise from the generated method code
- Parameters with default values in the C# interface will become optional? in the TypeScript interface

This C# code:

```
namespace SomeInterfaces
{
  public interface ISomething
  {
       string name {get;set;}
  }

  public interface IConsumer
  {
       string ost {get;set;}
       void consumeIt(string value, ISomething apa);
       string ReturnSomething(string value = null);
   }

```

will become this ts-code (with the returnPromise option set to true):

```
declare module MyModule {
    // models\test.cs
    export interface ISomething {
        name: string;
    }

    // models\test.cs
    export interface IConsumer {
        ost: string;
        consumeIt (value: string, apa: ISomething): Promise<void>
        ReturnSomething (value?: string): Promise<string>
    }

}
```

I hope you think this is an ok addition to your project and that I have not messed up your code to badly.

/David

### CODE REVIEW

**Comment by jonathanp** [comment]
Thanks for your PR @dlid! I'm glad that you like the project!

But I'm having trouble imagining a use case where you would need the C# methods in TypeScript interfaces? Could you describe the use case you're having please?

**Comment by dlid** [comment]
Sure thing. I can see why you ask, this is something we're attempting as an experient as well. But I'll explain how we're thinking.

We're refactoring an existing .NET Mvc solution (Front end and back end in the same application) where we split it into two separate projects. One pure front end solution using Angular 6 and TypeScript, and one .NET Core solution for the API.

The API project will have a C# interface defining the API-methods available, what types they return and the parameters. This is the interface that I'd like to automatically convert to typescript upon building the Angular application.

This way, if any method parameters have been changed or added in the .NET Core API project, we want the angular application build to fail so we can make sure the angular application is up to date with the API.

Basically it's an attempt to make the client side API calls type safe using the actual types from the API itself.

**Comment by jonathanp** [comment]
Ok, that's a cool experiment but I think it's a bit too specific to be a part of this module. But feel free to maintain your own fork, and I'll be sure to check it out if I even happen to be in a situation where that might be needed. 😊

**Comment by dlid** [comment]
Alright, no problem :) I'll maintain the fork and we'll see if we'll end up using it in our solution. 

Thanks!
